TestService_DownloadStateRunning fails intermittently

Jenkins reports that `TestService_DownloadStateRunning` fails from time to time.

And there seems also to be a race-condition while doing `TestClose`.

Still happening. Example:

```
=== RUN   TestService_DownloadStateRunning
1 20/27/02 13:43:26.182519347: server.go:217 (v3.(*Server).Start)       - Starting server at 2020-02-27 13:43:26 on address local://127.0.0.1:2000
2 20/27/02 13:43:26.182930432: websocket.go:186 (v3.(*WebSocket).start) - Starting to listen on 0.0.0.0:2001
1 20/27/02 13:43:26.339672741: server.go:217 (v3.(*Server).Start)       - Starting server at 2020-02-27 13:43:26 on address local://127.0.0.1:2010
2 20/27/02 13:43:26.340217772: websocket.go:186 (v3.(*WebSocket).start) - Starting to listen on 0.0.0.0:2011
1 20/27/02 13:43:26.507036294: server.go:217 (v3.(*Server).Start)       - Starting server at 2020-02-27 13:43:26 on address local://127.0.0.1:2020
2 20/27/02 13:43:26.507478118: websocket.go:186 (v3.(*WebSocket).start) - Starting to listen on 0.0.0.0:2021
1 20/27/02 13:43:26.667590053: server.go:217 (v3.(*Server).Start)       - Starting server at 2020-02-27 13:43:26 on address local://127.0.0.1:2030
2 20/27/02 13:43:26.668009898: websocket.go:186 (v3.(*WebSocket).start) - Starting to listen on 0.0.0.0:2031
2 20/27/02 13:43:26.780232500: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 0000000000000000000000000000000000000000000000000000000000000000 - contract: config
2 20/27/02 13:43:26.780536673: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1 - contract: darc
2 20/27/02 13:43:26.782606181: skipchain.go:212 (skipchain.(*Service).StoreSkipBlockInternal) - Creating new skipchain with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030]
2 20/27/02 13:43:26.788617559: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:26.800350738: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 0
2 20/27/02 13:43:26.802797379: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 0000000000000000000000000000000000000000000000000000000000000000 - contract: config
2 20/27/02 13:43:26.802932288: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1 - contract: darc
2 20/27/02 13:43:26.955807222: service.go:1640 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 new leader started polling for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:26.956887298: service.go:1665 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 starting heartbeat monitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with window 1h23m19.5s
2 20/27/02 13:43:26.957061997: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:26.958205686: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 48f6b6736dd58f83c260ca6d5c75381371f818f1e91ec27d35cfb1a74862a7a1
2 20/27/02 13:43:26.971808112: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:26.971959310: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:26.972259556: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.030074809: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 0
2 20/27/02 13:43:27.031805802: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 0
2 20/27/02 13:43:27.032756431: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 0
2 20/27/02 13:43:27.035326932: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 0000000000000000000000000000000000000000000000000000000000000000 - contract: config
2 20/27/02 13:43:27.035525467: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1 - contract: darc
2 20/27/02 13:43:27.037840787: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 0000000000000000000000000000000000000000000000000000000000000000 - contract: config
2 20/27/02 13:43:27.038141298: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1 - contract: darc
2 20/27/02 13:43:27.039293445: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 0000000000000000000000000000000000000000000000000000000000000000 - contract: config
2 20/27/02 13:43:27.039650529: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1 - contract: darc
2 20/27/02 13:43:27.142481151: service.go:1665 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 starting heartbeat monitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with window 1h23m19.5s
2 20/27/02 13:43:27.142686561: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.143453896: service.go:1665 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 starting heartbeat monitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with window 1h23m19.5s
2 20/27/02 13:43:27.143642452: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.143766958: service.go:1665 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 starting heartbeat monitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with window 1h23m19.5s
2 20/27/02 13:43:27.143949818: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.144308776: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 48f6b6736dd58f83c260ca6d5c75381371f818f1e91ec27d35cfb1a74862a7a1
2 20/27/02 13:43:27.145347322: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 48f6b6736dd58f83c260ca6d5c75381371f818f1e91ec27d35cfb1a74862a7a1
2 20/27/02 13:43:27.145639032: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 48f6b6736dd58f83c260ca6d5c75381371f818f1e91ec27d35cfb1a74862a7a1
1 20/27/02 13:43:27.191591167: service_test.go:2133 (byzcoin.TestService_DownloadStateRunning) - Adding dummy transactions
2 20/27/02 13:43:27.198870279: service.go:426 (byzcoin.(*Service).AddTransaction) - Instruction[0]: spawn:darc on instance ID 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1
2 20/27/02 13:43:27.243672990: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2134f79be7983721c7c5f44ce6b958e3f90f0e96feb92603d1ae4cfaa7a1719d - contract: darc
2 20/27/02 13:43:27.483757774: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2134f79be7983721c7c5f44ce6b958e3f90f0e96feb92603d1ae4cfaa7a1719d - contract: darc
2 20/27/02 13:43:27.485135274: skipchain.go:240 (skipchain.(*Service).StoreSkipBlockInternal) - Preparing to add block with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030] to b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.579909693: skipchain.go:1099 (skipchain.(*Service).forwardLinkLevel0) - local://127.0.0.1:2000 is adding forward-link level 0 to: 0->1 with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030]
2 20/27/02 13:43:27.585517129: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2000 asked to sign forward-link to block 1 : e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0 from b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.605753433: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2010 asked to sign forward-link to block 1 : e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0 from b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.607876107: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2020 asked to sign forward-link to block 1 : e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0 from b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.609064273: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2030 asked to sign forward-link to block 1 : e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0 from b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.621165617: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2134f79be7983721c7c5f44ce6b958e3f90f0e96feb92603d1ae4cfaa7a1719d - contract: darc
2 20/27/02 13:43:27.623260165: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2134f79be7983721c7c5f44ce6b958e3f90f0e96feb92603d1ae4cfaa7a1719d - contract: darc
2 20/27/02 13:43:27.623608186: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2134f79be7983721c7c5f44ce6b958e3f90f0e96feb92603d1ae4cfaa7a1719d - contract: darc
2 20/27/02 13:43:27.776793992: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 0: (0 height:[182 88 189 30 108 170 249 126 166 136 98 153 89 124 138 36 32 199 150 55 246 120 137 104 162 26 99 124 136 86 28 203] e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0)
2 20/27/02 13:43:27.776914004: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:27.777014846: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.794194398: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:27.801986432: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 1
2 20/27/02 13:43:27.860691209: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.861891754: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root c206e1550a12ef2538d4586670be0da1ec108110763e91dd416f49f7d7594779
2 20/27/02 13:43:27.861992912: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:27.867997223: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 0: (0 height:[182 88 189 30 108 170 249 126 166 136 98 153 89 124 138 36 32 199 150 55 246 120 137 104 162 26 99 124 136 86 28 203] e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0)
2 20/27/02 13:43:27.868114391: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:27.868205914: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.868491182: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 0: (0 height:[182 88 189 30 108 170 249 126 166 136 98 153 89 124 138 36 32 199 150 55 246 120 137 104 162 26 99 124 136 86 28 203] e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0)
2 20/27/02 13:43:27.868597211: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 0: (0 height:[182 88 189 30 108 170 249 126 166 136 98 153 89 124 138 36 32 199 150 55 246 120 137 104 162 26 99 124 136 86 28 203] e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0)
2 20/27/02 13:43:27.868744192: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:27.868848883: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:27.869066561: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.869187243: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:27.888354748: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:27.889792401: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:27.890405774: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:27.960734173: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 1
2 20/27/02 13:43:27.961783222: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 1
2 20/27/02 13:43:27.963720727: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 1
2 20/27/02 13:43:28.016499259: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.018211701: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root c206e1550a12ef2538d4586670be0da1ec108110763e91dd416f49f7d7594779
2 20/27/02 13:43:28.018282060: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.018779110: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.019133003: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.020918135: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root c206e1550a12ef2538d4586670be0da1ec108110763e91dd416f49f7d7594779
2 20/27/02 13:43:28.021037134: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.021222340: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root c206e1550a12ef2538d4586670be0da1ec108110763e91dd416f49f7d7594779
2 20/27/02 13:43:28.021390807: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.080896462: service.go:426 (byzcoin.(*Service).AddTransaction) - Instruction[0]: spawn:darc on instance ID 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1
2 20/27/02 13:43:28.325597069: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 14df992005b11b03fd94e350b5d7c72302a244b186d1b4a69e172995ab7de4c8 - contract: darc
2 20/27/02 13:43:28.482435420: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 14df992005b11b03fd94e350b5d7c72302a244b186d1b4a69e172995ab7de4c8 - contract: darc
2 20/27/02 13:43:28.483964364: skipchain.go:240 (skipchain.(*Service).StoreSkipBlockInternal) - Preparing to add block with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030] to b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.487725952: skipchain.go:1099 (skipchain.(*Service).forwardLinkLevel0) - local://127.0.0.1:2000 is adding forward-link level 0 to: 1->2 with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030]
2 20/27/02 13:43:28.494080918: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2000 asked to sign forward-link to block 2 : 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a from e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.508357406: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2030 asked to sign forward-link to block 2 : 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a from e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.509029739: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2020 asked to sign forward-link to block 2 : 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a from e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.510585222: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2010 asked to sign forward-link to block 2 : 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a from e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.525339126: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 14df992005b11b03fd94e350b5d7c72302a244b186d1b4a69e172995ab7de4c8 - contract: darc
2 20/27/02 13:43:28.526156976: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 14df992005b11b03fd94e350b5d7c72302a244b186d1b4a69e172995ab7de4c8 - contract: darc
2 20/27/02 13:43:28.526734644: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 14df992005b11b03fd94e350b5d7c72302a244b186d1b4a69e172995ab7de4c8 - contract: darc
2 20/27/02 13:43:28.672364132: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 1: (0 height:[229 86 251 170 100 163 11 118 190 4 255 30 26 58 25 129 25 111 15 98 252 123 171 1 228 174 89 142 17 202 34 160] 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a)
2 20/27/02 13:43:28.672557939: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:28.672679241: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.695903564: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:28.705583904: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 2
2 20/27/02 13:43:28.715800454: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.717214839: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 7474715c37cb077d7eecc4fb6e646c7f3e650ac8efa57ac503135f8e915b6b7b
2 20/27/02 13:43:28.717337226: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.725286213: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 1: (0 height:[229 86 251 170 100 163 11 118 190 4 255 30 26 58 25 129 25 111 15 98 252 123 171 1 228 174 89 142 17 202 34 160] 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a)
2 20/27/02 13:43:28.725586196: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:28.725699382: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 1: (0 height:[229 86 251 170 100 163 11 118 190 4 255 30 26 58 25 129 25 111 15 98 252 123 171 1 228 174 89 142 17 202 34 160] 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a)
2 20/27/02 13:43:28.725871890: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:28.726040592: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.726174308: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.726287243: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 1: (0 height:[229 86 251 170 100 163 11 118 190 4 255 30 26 58 25 129 25 111 15 98 252 123 171 1 228 174 89 142 17 202 34 160] 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a)
2 20/27/02 13:43:28.726373703: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:28.726499459: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:28.745590344: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:28.751974991: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:28.752222625: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:28.755402955: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 2
2 20/27/02 13:43:28.810912570: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 2
2 20/27/02 13:43:28.812299603: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 2
2 20/27/02 13:43:28.814329687: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.816030966: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 7474715c37cb077d7eecc4fb6e646c7f3e650ac8efa57ac503135f8e915b6b7b
2 20/27/02 13:43:28.816158160: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.862938748: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.863128774: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:28.864422404: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 7474715c37cb077d7eecc4fb6e646c7f3e650ac8efa57ac503135f8e915b6b7b
2 20/27/02 13:43:28.864492371: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.864618103: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 7474715c37cb077d7eecc4fb6e646c7f3e650ac8efa57ac503135f8e915b6b7b
2 20/27/02 13:43:28.864773610: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:28.940679609: service.go:426 (byzcoin.(*Service).AddTransaction) - Instruction[0]: spawn:darc on instance ID 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1
2 20/27/02 13:43:29.159359275: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 5a6d4da8225a1955087b451e59e016c2d90c9b3ddb1167a62eff5f64648882d7 - contract: darc
2 20/27/02 13:43:29.486124802: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 5a6d4da8225a1955087b451e59e016c2d90c9b3ddb1167a62eff5f64648882d7 - contract: darc
2 20/27/02 13:43:29.487620193: skipchain.go:240 (skipchain.(*Service).StoreSkipBlockInternal) - Preparing to add block with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030] to b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.491482920: skipchain.go:1099 (skipchain.(*Service).forwardLinkLevel0) - local://127.0.0.1:2000 is adding forward-link level 0 to: 2->3 with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030]
2 20/27/02 13:43:29.496312067: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2000 asked to sign forward-link to block 3 : 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411 from 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.509557325: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2020 asked to sign forward-link to block 3 : 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411 from 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.510696816: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2030 asked to sign forward-link to block 3 : 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411 from 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.513887957: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2010 asked to sign forward-link to block 3 : 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411 from 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.529353104: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 5a6d4da8225a1955087b451e59e016c2d90c9b3ddb1167a62eff5f64648882d7 - contract: darc
2 20/27/02 13:43:29.532421008: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 5a6d4da8225a1955087b451e59e016c2d90c9b3ddb1167a62eff5f64648882d7 - contract: darc
2 20/27/02 13:43:29.535154792: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 5a6d4da8225a1955087b451e59e016c2d90c9b3ddb1167a62eff5f64648882d7 - contract: darc
2 20/27/02 13:43:29.680224170: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 2: (0 height:[49 105 35 20 218 24 87 149 218 166 5 178 245 142 61 4 94 28 21 90 34 111 222 250 162 177 118 217 111 102 81 154] 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411)
2 20/27/02 13:43:29.680335706: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:29.680429132: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.698751112: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:29.708311043: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 3
2 20/27/02 13:43:29.719018188: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.720558537: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 8bf4e0a455e6557292fcdd3a8c779ac172eb659769a0e475d345f425c312c1a9
2 20/27/02 13:43:29.720639979: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:29.729428058: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 2: (0 height:[49 105 35 20 218 24 87 149 218 166 5 178 245 142 61 4 94 28 21 90 34 111 222 250 162 177 118 217 111 102 81 154] 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411)
2 20/27/02 13:43:29.729685851: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:29.729828505: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 2: (0 height:[49 105 35 20 218 24 87 149 218 166 5 178 245 142 61 4 94 28 21 90 34 111 222 250 162 177 118 217 111 102 81 154] 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411)
2 20/27/02 13:43:29.730006003: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:29.730136214: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.730312298: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 2: (0 height:[49 105 35 20 218 24 87 149 218 166 5 178 245 142 61 4 94 28 21 90 34 111 222 250 162 177 118 217 111 102 81 154] 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411)
2 20/27/02 13:43:29.730456696: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.730585754: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:29.730766808: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:29.749547897: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:29.750484077: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:29.750592078: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:29.812403802: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 3
2 20/27/02 13:43:29.812768032: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 3
2 20/27/02 13:43:29.812902414: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 3
2 20/27/02 13:43:29.877105183: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.877575299: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.877825669: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.879470655: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 8bf4e0a455e6557292fcdd3a8c779ac172eb659769a0e475d345f425c312c1a9
2 20/27/02 13:43:29.879571209: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:29.879886333: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 8bf4e0a455e6557292fcdd3a8c779ac172eb659769a0e475d345f425c312c1a9
2 20/27/02 13:43:29.880000472: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:29.880344507: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 8bf4e0a455e6557292fcdd3a8c779ac172eb659769a0e475d345f425c312c1a9
2 20/27/02 13:43:29.880515552: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
1 20/27/02 13:43:29.933595524: service_test.go:2139 (byzcoin.TestService_DownloadStateRunning) - Not deleting leader
1 20/27/02 13:43:29.933774239: service_test.go:2142 (byzcoin.TestService_DownloadStateRunning) - Deleting node 1 and adding new transaction
1 20/27/02 13:43:29.933864073: service_test.go:2708 (byzcoin.(*ser).deleteDBs) - Deleting DB of node 1
1 20/27/02 13:43:29.934034558: service.go:2501 (byzcoin.(*Service).cleanupGoroutines) - local://127.0.0.1:2010 closing go-routines
2 20/27/02 13:43:29.934444980: service.go:2560 (byzcoin.(*Service).monitorLeaderFailure) - local://127.0.0.1:2010 closing heartbeat timeout monitor
2 20/27/02 13:43:29.959602859: service.go:1343 (byzcoin.(*Service).catchupFromID) - got asked for an unknown, friendly skipchain: b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
1 20/27/02 13:43:29.959852830: service.go:1357 (byzcoin.(*Service).catchupFromID) - local://127.0.0.1:2010: catching up with chain b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:29.962830953: websocket.go:280 (v3.wsHandler.ServeHTTP) - ws request from 127.0.0.1:45590: Skipchain/GetSingleBlock
2 20/27/02 13:43:29.963055275: websocket.go:280 (v3.wsHandler.ServeHTTP) - ws request from 127.0.0.1:57484: Skipchain/GetSingleBlock
2 20/27/02 13:43:29.968902537: websocket.go:245 (v3.wsHandler.ServeHTTP.func1) - ws close 127.0.0.1:45590 n 1 rx 34 tx 3134
2 20/27/02 13:43:29.969264751: websocket.go:245 (v3.wsHandler.ServeHTTP.func1) - ws close 127.0.0.1:57484 n 1 rx 34 tx 3134
1 20/27/02 13:43:29.970673707: service.go:1385 (byzcoin.(*Service).catchUp) - local://127.0.0.1:2010 Catching up b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb / 3
2 20/27/02 13:43:29.971471413: service.go:426 (byzcoin.(*Service).AddTransaction) - Instruction[0]: spawn:darc on instance ID 45272504b9d7cc6a9885b8998942ee9441a03ec258dbfbcc20d6cd610b7817e1
2 20/27/02 13:43:29.973424667: service.go:1431 (byzcoin.(*Service).catchUp) - local://127.0.0.1:2010 Downloading whole DB for catching up
2 20/27/02 13:43:29.973538952: service.go:1149 (byzcoin.(*Service).downloadDB) - local://127.0.0.1:2010: downloading DB
2 20/27/02 13:43:29.976074886: websocket.go:280 (v3.wsHandler.ServeHTTP) - ws request from 127.0.0.1:45592: ByzCoin/DownloadState
2 20/27/02 13:43:29.976244179: service.go:617 (byzcoin.(*Service).DownloadState) - local://127.0.0.1:2000 Creating new download
2 20/27/02 13:43:30.082171847: websocket.go:245 (v3.wsHandler.ServeHTTP.func1) - ws close 127.0.0.1:45592 n 1 rx 39 tx 6395
1 20/27/02 13:43:30.082508910: service.go:1185 (byzcoin.(*Service).downloadDB.func1) - Downloaded key/values 0..17 of 17 from local://127.0.0.1:2000
2 20/27/02 13:43:30.089376582: websocket.go:280 (v3.wsHandler.ServeHTTP) - ws request from 127.0.0.1:60638: Skipchain/GetUpdateChain
2 20/27/02 13:43:30.089703289: websocket.go:280 (v3.wsHandler.ServeHTTP) - ws request from 127.0.0.1:57492: Skipchain/GetUpdateChain
2 20/27/02 13:43:30.102473619: websocket.go:245 (v3.wsHandler.ServeHTTP.func1) - ws close 127.0.0.1:60638 n 1 rx 38 tx 16878
2 20/27/02 13:43:30.102695680: websocket.go:245 (v3.wsHandler.ServeHTTP.func1) - ws close 127.0.0.1:57492 n 1 rx 38 tx 16878
2 20/27/02 13:43:30.158485142: service.go:1245 (byzcoin.(*Service).downloadDB.func1) - Storing block 0: b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:30.158628486: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 0 / b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:30.179823433: service.go:1245 (byzcoin.(*Service).downloadDB.func1) - Storing block 1: e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:30.179907224: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 1 / e556fbaa64a30b76be04ff1e1a3a1981196f0f62fc7bab01e4ae598e11ca22a0
2 20/27/02 13:43:30.202004845: service.go:1245 (byzcoin.(*Service).downloadDB.func1) - Storing block 2: 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:30.202088161: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 2 / 31692314da185795daa605b2f58e3d045e1c155a226fdefaa2b176d96f66519a
2 20/27/02 13:43:30.205864249: tx_pipeline.go:187 (byzcoin.(*defaultTxProcessor).CollectTx) - local://127.0.0.1:2000 timeout while collecting transactions from other nodes
2 20/27/02 13:43:30.225349956: service.go:1245 (byzcoin.(*Service).downloadDB.func1) - Storing block 3: 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:30.225465511: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
1 20/27/02 13:43:30.231609546: service.go:1248 (byzcoin.(*Service).downloadDB.func1) - local://127.0.0.1:2010: successfully downloaded database for chain b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb up to block 3/3
2 20/27/02 13:43:30.495515356: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2f17e6318b60f1a29bd630fd523fac887354ffd9f3ea32ab8789541e9ef2c9f3 - contract: darc
2 20/27/02 13:43:30.999901066: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2f17e6318b60f1a29bd630fd523fac887354ffd9f3ea32ab8789541e9ef2c9f3 - contract: darc
2 20/27/02 13:43:31.001291027: skipchain.go:240 (skipchain.(*Service).StoreSkipBlockInternal) - Preparing to add block with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030] to b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:31.007379566: skipchain.go:1099 (skipchain.(*Service).forwardLinkLevel0) - local://127.0.0.1:2000 is adding forward-link level 0 to: 3->4 with roster [local://127.0.0.1:2000 local://127.0.0.1:2010 local://127.0.0.1:2020 local://127.0.0.1:2030]
2 20/27/02 13:43:31.015197075: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2000 asked to sign forward-link to block 4 : b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f from 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.036331730: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2020 asked to sign forward-link to block 4 : b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f from 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.037161177: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2010 asked to sign forward-link to block 4 : b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f from 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.037385173: skipchain.go:1177 (skipchain.(*Service).bftForwardLinkLevel0) - local://127.0.0.1:2030 asked to sign forward-link to block 4 : b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f from 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.055531216: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2f17e6318b60f1a29bd630fd523fac887354ffd9f3ea32ab8789541e9ef2c9f3 - contract: darc
2 20/27/02 13:43:31.060448447: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2f17e6318b60f1a29bd630fd523fac887354ffd9f3ea32ab8789541e9ef2c9f3 - contract: darc
2 20/27/02 13:43:31.060637683: service.go:2251 (byzcoin.(*Service).processOneTx) - StateChange Create for id 2f17e6318b60f1a29bd630fd523fac887354ffd9f3ea32ab8789541e9ef2c9f3 - contract: darc
2 20/27/02 13:43:31.195163151: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 3: (0 height:[104 128 240 217 145 183 227 184 181 55 4 146 114 165 115 120 62 185 142 132 8 205 204 202 143 79 245 131 9 65 180 17] b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f)
2 20/27/02 13:43:31.195341456: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:31.195457272: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.221812395: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 4 / b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f
2 20/27/02 13:43:31.231525196: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 4
2 20/27/02 13:43:31.241090691: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:31.242379592: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2000 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 83feee9fa35ce9a0015e0db82c2c0079c266172040286bad0bc3f77dca82951d
2 20/27/02 13:43:31.242436603: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:31.247735193: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 3: (0 height:[104 128 240 217 145 183 227 184 181 55 4 146 114 165 115 120 62 185 142 132 8 205 204 202 143 79 245 131 9 65 180 17] b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f)
2 20/27/02 13:43:31.247891937: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:31.248158862: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.248284952: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 3: (0 height:[104 128 240 217 145 183 227 184 181 55 4 146 114 165 115 120 62 185 142 132 8 205 204 202 143 79 245 131 9 65 180 17] b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f)
2 20/27/02 13:43:31.248387200: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:31.248501664: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.248711691: skipchain.go:1585 (skipchain.(*Service).propagateForwardLinkHandler) - Adding Forwardlink to block 3: (0 height:[104 128 240 217 145 183 227 184 181 55 4 146 114 165 115 120 62 185 142 132 8 205 204 202 143 79 245 131 9 65 180 17] b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f)
2 20/27/02 13:43:31.248837623: skipchain.go:1613 (skipchain.(*Service).propagateForwardLinkHandler) - Storing new forward-link and eventual new block
2 20/27/02 13:43:31.248980596: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 3 / 6880f0d991b7e3b8b537049272a573783eb98e8408cdccca8f4ff5830941b411
2 20/27/02 13:43:31.268041002: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 4 / b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f
2 20/27/02 13:43:31.273118030: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 4 / b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f
2 20/27/02 13:43:31.273723108: struct.go:798 (skipchain.(*SkipBlockDB).StoreBlocks.func1) - Storing skipblock 4 / b3ddf8b611b74fc20b2c6dca5a5e2ed026979441826b42554394a9acd6096e0f
2 20/27/02 13:43:31.279873314: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 4
2 20/27/02 13:43:31.338979631: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 4
2 20/27/02 13:43:31.339562740: service.go:1588 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 Updating 1 transactions for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb on index 4
2 20/27/02 13:43:31.342027715: service.go:1665 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 starting heartbeat monitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with window 1h23m19.5s
2 20/27/02 13:43:31.342223064: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:31.343541556: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2010 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 83feee9fa35ce9a0015e0db82c2c0079c266172040286bad0bc3f77dca82951d
2 20/27/02 13:43:31.343642295: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:31.357095388: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:31.357361290: service.go:1683 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 started viewchangeMonitor for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb
2 20/27/02 13:43:31.359265616: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2020 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 83feee9fa35ce9a0015e0db82c2c0079c266172040286bad0bc3f77dca82951d
2 20/27/02 13:43:31.359389487: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
2 20/27/02 13:43:31.359685959: service.go:1704 (byzcoin.(*Service).updateTrieCallback) - local://127.0.0.1:2030 updated trie for b658bd1e6caaf97ea6886299597c8a2420c79637f6788968a21a637c88561ccb with root 83feee9fa35ce9a0015e0db82c2c0079c266172040286bad0bc3f77dca82951d
2 20/27/02 13:43:31.359952864: skipchain.go:1606 (skipchain.(*Service).propagateForwardLinkHandler.func1) - Clearing block
1 20/27/02 13:43:31.371455172: service_test.go:2142 (byzcoin.TestService_DownloadStateRunning) - Deleting node 2 and adding new transaction
1 20/27/02 13:43:31.371602609: service_test.go:2708 (byzcoin.(*ser).deleteDBs) - Deleting DB of node 2
1 20/27/02 13:43:31.371721679: service.go:2501 (byzcoin.(*Service).cleanupGoroutines) - local://127.0.0.1:2020 closing go-routines
2 20/27/02 13:43:31.372030724: service.go:2560 (byzcoin.(*Service).monitorLeaderFailure) - local://127.0.0.1:2020 closing heartbeat timeout monitor
2 20/27/02 13:43:31.374495379: skipchain.go:382 (skipchain.(*Service).StoreSkipBlockInternal) - local://127.0.0.1:2000: sending request for height 1 to local://127.0.0.1:2000
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9240c8]

goroutine 38283 [running]:
go.etcd.io/bbolt.(*Bucket).Cursor(...)
	/go/pkg/mod/go.etcd.io/bbolt@v1.3.3/bucket.go:84
go.etcd.io/bbolt.(*Bucket).Get(0x0, 0xc004b72d99, 0x20, 0x20, 0x0, 0x92f790, 0xc0012477d8)
	/go/pkg/mod/go.etcd.io/bbolt@v1.3.3/bucket.go:260 +0x48
go.dedis.ch/cothority/v3/skipchain.(*SkipBlockDB).getFromTx(0xc004632980, 0xc000255260, 0xc004b72d99, 0x20, 0x20, 0x92f472, 0xc001247600, 0xc000255260)
	/cothority/skipchain/struct.go:1287 +0xc6
go.dedis.ch/cothority/v3/skipchain.(*SkipBlockDB).GetByID.func1(0xc000255260, 0x0, 0xc000255260)
	/cothority/skipchain/struct.go:778 +0x74
go.etcd.io/bbolt.(*DB).View(0xc001247600, 0xc0024f3ad8, 0x0, 0x0)
	/go/pkg/mod/go.etcd.io/bbolt@v1.3.3/db.go:725 +0xe0
go.dedis.ch/cothority/v3/skipchain.(*SkipBlockDB).GetByID(0xc004632980, 0xc004b72d99, 0x20, 0x20, 0x20)
	/cothority/skipchain/struct.go:777 +0xce
go.dedis.ch/cothority/v3/skipchain.(*Service).bftForwardLink.func1(0xc004b72d36, 0xf4f, 0xf4f, 0xc0009073b0, 0xc004b72d13, 0x20, 0x20, 0x0, 0x0)
	/cothority/skipchain/skipchain.go:1404 +0x3ba
go.dedis.ch/cothority/v3/skipchain.(*Service).bftForwardLink(0xc0009073b0, 0xc004b72d13, 0x20, 0x20, 0xc004b72d36, 0xf4f, 0xf4f, 0x0)
	/cothority/skipchain/skipchain.go:1459 +0x9f
go.dedis.ch/cothority/v3/blscosi/protocol.(*SubBlsCosi).makeVerification(0xc0003b8180, 0xc000693080)
	/cothority/blscosi/protocol/sub_protocol.go:399 +0xd2
created by go.dedis.ch/cothority/v3/blscosi/protocol.(*SubBlsCosi).dispatchLeaf
	/cothority/blscosi/protocol/sub_protocol.go:339 +0x134
FAIL	go.dedis.ch/cothority/v3/byzcoin	383.527s
FAIL
Build step 'Execute shell' marked build as failure
An attempt to send an e-mail to empty list of recipients, ignored.
[Slack Notifications] found #4715 as previous completed, non-aborted build
[Slack Notifications] will send OnEveryFailureNotification because build matches and user preferences allow it
Finished: FAILURE
```

--- a/byzcoin/api_test.go
+++ b/byzcoin/api_test.go
@@ -307,6 +307,7 @@ func TestClient_NoPhantomSkipchain(t *testing.T) {
 	gac, err = c.GetAllByzCoinIDs(roster.List[0])
 	require.NoError(t, err)
 	require.Equal(t, 1, len(gac.IDs))
+	require.NoError(t, l.WaitDone(time.Second))
 }
 
 // Insure that the decoder will return an error if the reply

--- a/byzcoin/service.go
+++ b/byzcoin/service.go
@@ -2567,7 +2567,6 @@ func (s *Service) TestRestart() error {
 }
 
 func (s *Service) cleanupGoroutines() {
-	log.Lvl1(s.ServerIdentity(), "closing go-routines 1")
 	s.heartbeats.closeAll()
 	s.closeLeaderMonitorChan <- true
 	s.viewChangeMan.closeAll()
@@ -2579,9 +2578,7 @@ func (s *Service) cleanupGoroutines() {
 		delete(s.pollChan, k)
 	}
 	s.pollChanMut.Unlock()
-	log.Lvl1(s.ServerIdentity(), "closing go-routines 2")
 	s.pollChanWG.Wait()
-	log.Lvl1(s.ServerIdentity(), "closing go-routines 3")
 }
 
 func (s *Service) monitorLeaderFailure() {

--- a/byzcoin/service.go
+++ b/byzcoin/service.go
@@ -15,6 +15,8 @@ import (
 	"sync"
 	"time"
 
+	uuid "gopkg.in/satori/go.uuid.v1"
+
 	"go.dedis.ch/cothority/v3"
 	"go.dedis.ch/cothority/v3/blscosi/protocol"
 	"go.dedis.ch/cothority/v3/byzcoin/trie"
@@ -31,7 +33,6 @@ import (
 	"go.dedis.ch/protobuf"
 	"go.etcd.io/bbolt"
 	"golang.org/x/xerrors"
-	uuid "gopkg.in/satori/go.uuid.v1"
 )
 
 var pairingSuite = suites.MustFind("bn256.adapter").(*pairing.SuiteBn256)
@@ -2492,8 +2493,7 @@ func (s *Service) getTxs(leader *network.ServerIdentity, roster *onet.Roster, sc
 
 	// First we check if we are up-to-date with this chain and catch up
 	// if necessary.
-	_, doCatchUp := s.skService().WaitBlock(scID, latestID)
-	if doCatchUp {
+	if !s.skService().ChainHasBlock(scID, latestID) {
 		// The function will prevent multiple request to catch up so we can securely call it here
 		err := s.catchupFromID(roster, scID, latestID)
 		if err != nil {
@@ -2547,6 +2547,7 @@ func loadNonceFromTxs(txs TxResults) ([]byte, error) {
 func (s *Service) TestClose() {
 	s.closedMutex.Lock()
 	if !s.closed {
+		s.skService().TestClose()
 		s.closed = true
 		s.closedMutex.Unlock()
 		s.cleanupGoroutines()
@@ -2556,8 +2557,17 @@ func (s *Service) TestClose() {
 	}
 }
 
+// TestRestart activates a test that has been closed using TestClose. This
+// allows to simulate restarting of nodes in the tests.
+func (s *Service) TestRestart() error {
+	if err := s.startAllChains(); err != nil {
+		return err
+	}
+	return s.skService().TestRestart()
+}
+
 func (s *Service) cleanupGoroutines() {
-	log.Lvl1(s.ServerIdentity(), "closing go-routines")
+	log.Lvl1(s.ServerIdentity(), "closing go-routines 1")
 	s.heartbeats.closeAll()
 	s.closeLeaderMonitorChan <- true
 	s.viewChangeMan.closeAll()
@@ -2569,7 +2579,9 @@ func (s *Service) cleanupGoroutines() {
 		delete(s.pollChan, k)
 	}
 	s.pollChanMut.Unlock()
+	log.Lvl1(s.ServerIdentity(), "closing go-routines 2")
 	s.pollChanWG.Wait()
+	log.Lvl1(s.ServerIdentity(), "closing go-routines 3")
 }
 
 func (s *Service) monitorLeaderFailure() {

--- a/byzcoin/service_test.go
+++ b/byzcoin/service_test.go
@@ -1028,7 +1028,8 @@ func sendTransactionWithCounter(t *testing.T, s *ser, client int, kind string, w
 		InclusionWait: wait,
 	})
 
-	for doCatchUp := false; !doCatchUp && wait != 0; _, doCatchUp = s.services[client].skService().WaitBlock(s.genesis.SkipChainID(), nil) {
+	for isProcessing := true; isProcessing && wait != 0; {
+		isProcessing = ser.skService().ChainIsProcessing(s.genesis.SkipChainID())
 		time.Sleep(s.interval)
 	}
 
@@ -2709,8 +2710,8 @@ func (s *ser) testDarcEvolution(t *testing.T, d2 darc.Darc, fail bool) (pr *Proo
 }
 
 func (s *ser) deleteDBs(t *testing.T, index int) {
-	log.Lvl1("Deleting DB of node", index)
 	bc := s.services[index]
+	log.Lvl1("Deleting DB of node", index, bc.ServerIdentity())
 	bc.TestClose()
 	for scid := range bc.stateTries {
 		require.NoError(t, deleteDB(bc.ServiceProcessor, []byte(scid)))
@@ -2721,8 +2722,7 @@ func (s *ser) deleteDBs(t *testing.T, index int) {
 	sc := bc.Service(skipchain.ServiceName).(*skipchain.Service)
 	require.NoError(t, deleteDB(sc.ServiceProcessor, []byte("skipblocks")))
 	require.NoError(t, deleteDB(sc.ServiceProcessor, []byte("skipchainconfig")))
-	sc.TestRestart()
-	require.NoError(t, bc.startAllChains())
+	require.NoError(t, bc.TestRestart())
 }
 
 // Waits to have a coherent view in all nodes with at least the block

--- a/byzcoin/service_test.go
+++ b/byzcoin/service_test.go
@@ -2711,7 +2711,7 @@ func (s *ser) testDarcEvolution(t *testing.T, d2 darc.Darc, fail bool) (pr *Proo
 
 func (s *ser) deleteDBs(t *testing.T, index int) {
 	bc := s.services[index]
-	log.Lvl1("Deleting DB of node", index, bc.ServerIdentity())
+	log.Lvlf1("%s: Deleting DB of node %d", bc.ServerIdentity(), index)
 	bc.TestClose()
 	for scid := range bc.stateTries {
 		require.NoError(t, deleteDB(bc.ServiceProcessor, []byte(scid)))

--- a/byzcoin/tx_pipeline.go
+++ b/byzcoin/tx_pipeline.go
@@ -110,7 +110,11 @@ func (s *defaultTxProcessor) CollectTx() (*collectTxResult, error) {
 		return nil, xerrors.Errorf("reading config: %v", err)
 	}
 
-	_, isNotProcessingBlock := s.skService().WaitBlock(s.scID, nil)
+	if s.skService().ChainIsProcessing(s.scID) {
+		// When a block is processed, wait for the block to be done
+		time.Sleep(bcConfig.BlockInterval / 2)
+		return &collectTxResult{Txs: nil, CommonVersion: CurrentVersion - 1}, nil
+	}
 
 	latest, err := s.db().GetLatestByID(s.scID)
 	if err != nil {
@@ -140,11 +144,6 @@ func (s *defaultTxProcessor) CollectTx() (*collectTxResult, error) {
 	root := proto.(*CollectTxProtocol)
 	root.SkipchainID = s.scID
 	root.LatestID = latest.Hash
-	// When a block is processed, we prevent conodes to send us back transactions
-	// until the next collection.
-	if !isNotProcessingBlock {
-		root.MaxNumTxs = 0
-	}
 
 	log.Lvl3("Asking", root.Roster().List, "for Txs")
 	if err := root.Start(); err != nil {

--- a/byzcoin/tx_pipeline.go
+++ b/byzcoin/tx_pipeline.go
@@ -111,8 +111,8 @@ func (s *defaultTxProcessor) CollectTx() (*collectTxResult, error) {
 	}
 
 	if s.skService().ChainIsProcessing(s.scID) {
-		// When a block is processed, wait for the block to be done
-		time.Sleep(bcConfig.BlockInterval / 2)
+		// When a block is processed,
+		// return immediately without processing any tx from the nodes.
 		return &collectTxResult{Txs: nil, CommonVersion: CurrentVersion - 1}, nil
 	}
 

--- a/go.mod
+++ b/go.mod
@@ -7,7 +7,7 @@ require (
 	github.com/allegro/bigcache v1.2.1 // indirect
 	github.com/aristanetworks/goarista v0.0.0-20191023202215-f096da5361bb // indirect
 	github.com/bford/golang-x-crypto v0.0.0-20160518072526-27db609c9d03
-	github.com/btcsuite/btcd v0.20.0-beta // indirect
+	github.com/btcsuite/btcd v0.20.1-beta // indirect
 	github.com/coreos/go-oidc v2.2.1+incompatible
 	github.com/cpuguy83/go-md2man/v2 v2.0.0 // indirect
 	github.com/deckarep/golang-set v1.7.1 // indirect

--- a/skipchain/skipchain.go
+++ b/skipchain/skipchain.go
@@ -974,7 +974,8 @@ func (s *Service) ChainIsProcessing(sid SkipBlockID) bool {
 // WaitBlock returns a block by its ID instantly if already stored in the DB
 // or check if the block is inside the buffer. If the block is known, false will
 // be returned because a catch up is not necessary, true otherwise.
-// DEPRECATED
+// DEPRECATED - mixes checking for available block and waiting for it.
+// Use ChainHasBlock and ChainIsProcessing instead.
 func (s *Service) WaitBlock(sid SkipBlockID, id SkipBlockID) (*SkipBlock, bool) {
 	sb := s.db.GetByID(id)
 	if sb != nil {

--- a/skipchain/skipchain.go
+++ b/skipchain/skipchain.go
@@ -957,9 +957,24 @@ func (s *Service) ListFollow(list *ListFollow) (*ListFollowReply, error) {
 	return reply, nil
 }
 
+// ChainHasBlock returns true if the given chain has the block id, either in
+// the database or in the blockBuffer. In the latter case, it means that the
+// block has been proposed and is waiting to be accepted.
+func (s *Service) ChainHasBlock(sid SkipBlockID, id SkipBlockID) bool {
+	return s.db.GetByID(id) != nil ||
+		s.blockBuffer.get(sid, id) != nil
+}
+
+// ChainIsProcessing returns true if the given chain has a block in the
+// blockBuffer.
+func (s *Service) ChainIsProcessing(sid SkipBlockID) bool {
+	return s.blockBuffer.has(sid)
+}
+
 // WaitBlock returns a block by its ID instantly if already stored in the DB
 // or check if the block is inside the buffer. If the block is known, false will
 // be returned because a catch up is not necessary, true otherwise.
+// DEPRECATED
 func (s *Service) WaitBlock(sid SkipBlockID, id SkipBlockID) (*SkipBlock, bool) {
 	sb := s.db.GetByID(id)
 	if sb != nil {
@@ -982,6 +997,11 @@ func (s *Service) GetDB() *SkipBlockDB {
 // NewProtocol intercepts the creation of the skipblock protocol and
 // initialises the necessary variables.
 func (s *Service) NewProtocol(ti *onet.TreeNodeInstance, conf *onet.GenericConfig) (pi onet.ProtocolInstance, err error) {
+	s.closedMutex.Lock()
+	defer s.closedMutex.Unlock()
+	if s.closed {
+		return nil, xerrors.New("don't process protocols if closed")
+	}
 	if ti.ProtocolName() == ProtocolExtendRoster {
 		// Start by getting latest blocks of all followers
 		pi, err = NewProtocolExtendRoster(ti)
@@ -1055,8 +1075,10 @@ func (s *Service) TestRestart() error {
 	//s.verifiers = map[VerifierID]SkipBlockVerifier{}
 	s.propTimeout = defaultPropagateTimeout
 	s.blockBuffer = newSkipBlockBuffer()
+	s.closedMutex.Lock()
 	s.closed = false
 	s.closing = make(chan bool)
+	s.closedMutex.Unlock()
 	return s.tryLoad()
 }
 

--- a/skipchain/skipchain_test.go
+++ b/skipchain/skipchain_test.go
@@ -1436,36 +1436,37 @@ func TestRosterAddCausesSync(t *testing.T) {
 	}
 }
 
-func TestService_WaitBlock(t *testing.T) {
+func TestService_ChainBlock(t *testing.T) {
 	local := onet.NewLocalTest(cothority.Suite)
 	defer local.CloseAll()
 	_, ro, service := local.MakeSRS(cothority.Suite, 1, skipchainSID)
 
 	s := service.(*Service)
 
-	sb, doCatchUp := s.WaitBlock(SkipBlockID{}, nil)
-	require.Nil(t, sb)
-	require.True(t, doCatchUp)
+	isProcessing := s.ChainIsProcessing(SkipBlockID{})
+	require.False(t, isProcessing)
+	hasBlock := s.ChainHasBlock(SkipBlockID{}, SkipBlockID{})
+	require.False(t, hasBlock)
 
 	sbRoot, err := makeGenesisRoster(s, ro)
 	require.NoError(t, err)
 
-	sb, doCatchUp = s.WaitBlock(sbRoot.Hash, sbRoot.Hash)
-	require.NotNil(t, sb)
-	require.False(t, doCatchUp)
-
-	sb, doCatchUp = s.WaitBlock(sbRoot.Hash, SkipBlockID{})
-	require.Nil(t, sb)
-	require.True(t, doCatchUp)
+	isProcessing = s.ChainIsProcessing(sbRoot.Hash)
+	require.False(t, isProcessing)
+	hasBlock = s.ChainHasBlock(sbRoot.Hash, sbRoot.Hash)
+	require.True(t, hasBlock)
+	hasBlock = s.ChainHasBlock(sbRoot.Hash, SkipBlockID{})
+	require.False(t, hasBlock)
 
 	newSb := NewSkipBlock()
 	newSb.Index = 1
 	newSb.GenesisID = sbRoot.Hash
 	newSb.updateHash()
 	s.blockBuffer.add(newSb)
-	sb, doCatchUp = s.WaitBlock(sbRoot.Hash, newSb.Hash)
-	require.Nil(t, sb)
-	require.False(t, doCatchUp)
+	isProcessing = s.ChainIsProcessing(sbRoot.Hash)
+	require.True(t, isProcessing)
+	hasBlock = s.ChainHasBlock(sbRoot.Hash, newSb.Hash)
+	require.True(t, hasBlock)
 }
 
 // The test scenario is the following:
@@ -1537,6 +1538,8 @@ func TestService_MissingForwardLinks(t *testing.T) {
 	require.NoError(t, err)
 	// C4 is re-inserted and should then update the missing forward link with the new block.
 	require.Equal(t, 2, len(sbAtIndex2FromC4.SkipBlock.ForwardLink))
+
+	require.NoError(t, local.WaitDone(time.Second))
 }
 
 func TestService_ForwardLinkVerification(t *testing.T) {
@@ -1594,6 +1597,8 @@ func TestService_ForwardLinkVerification(t *testing.T) {
 	fs.TargetHeight = 0
 	require.False(t, s.bftForwardLink([]byte{}, marshal(fs)))
 	require.Contains(t, log.GetStdErr(), "target height does not match")
+
+	require.NoError(t, local.WaitDone(time.Second))
 }
 
 // Test if the optimization works as expected for a normal situation.

--- a/skipchain/struct.go
+++ b/skipchain/struct.go
@@ -20,6 +20,8 @@ import (
 	"go.dedis.ch/onet/v3"
 	"go.dedis.ch/onet/v3/log"
 	"go.dedis.ch/onet/v3/network"
+	"golang.org/x/xerrors"
+
 	bbolt "go.etcd.io/bbolt"
 	uuid "gopkg.in/satori/go.uuid.v1"
 )
@@ -774,6 +776,9 @@ func (db *SkipBlockDB) GetStatus() *onet.Status {
 // GetByID returns a new copy of the skip-block or nil if it doesn't exist
 func (db *SkipBlockDB) GetByID(sbID SkipBlockID) *SkipBlock {
 	var result *SkipBlock
+	if sbID == nil {
+		return nil
+	}
 	err := db.View(func(tx *bbolt.Tx) error {
 		sb, err := db.getFromTx(tx, sbID)
 		if err != nil {
@@ -1284,7 +1289,11 @@ func (db *SkipBlockDB) storeToTx(tx *bbolt.Tx, sb *SkipBlock) error {
 // An error is thrown if marshalling fails.
 // The caller must ensure that this function is called from within a valid transaction.
 func (db *SkipBlockDB) getFromTx(tx *bbolt.Tx, sbID SkipBlockID) (*SkipBlock, error) {
-	val := tx.Bucket([]byte(db.bucketName)).Get(sbID)
+	if sbID == nil {
+		return nil, xerrors.New("cannot look up skipblock with ID == nil")
+	}
+
+	val := tx.Bucket(db.bucketName).Get(sbID)
 	if val == nil {
 		return nil, nil
 	}

--- a/skipchain/struct_test.go
+++ b/skipchain/struct_test.go
@@ -6,6 +6,7 @@ import (
 	"io/ioutil"
 	"os"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -139,6 +140,8 @@ func TestSkipBlock_InvalidForwardLinks(t *testing.T) {
 	// Try to store a new block with too many forward-links
 	s.db.Store(gb2)
 	require.Contains(t, log.GetStdErr(), "found 1 forward-links for a height of 0")
+
+	require.NoError(t, local.WaitDone(time.Second))
 }
 
 func TestSkipBlock_WrongSignatures(t *testing.T) {

